benchmark_runner/workloads/bootstorm_vm.py

```diff
@@ -173,99 +173,132 @@ def _verify_single_vm(self, vm_name, retries=5, delay=10):
-                elif self._endpoint_url and not self._google_drive_shared_drive_id:
-                    self.upload_run_artifacts_to_s3()
-                else:
-                    self._save_artifacts_local = True
+    def _verify_vms_in_parallel(self, vm_names):
+        """
+        This method verifies vms in parallel
+        :param vm_names:
+        :return:
+        """
+        failure_vms = []
+        manager = Manager()
+        return_dict = manager.dict()
+
+        # Split vm_names according to self._threads_limit
+        for i in range(0, len(vm_names), self._threads_limit):
+            bulk = vm_names[i:i + self._threads_limit]
+            processes = []
+
+            for vm_name in bulk:
+                p = Process(target=self._verify_single_vm_wrapper, args=(vm_name, return_dict))
+                p.start()
+                processes.append(p)
+
+            for p in processes:
+                p.join()
+
+        # After all processes are done, collect failures
+        for vm_name, status in return_dict.items():
+            if status.lower() != 'true':
+                failure_vms.append(vm_name)
+        return failure_vms
+
+    def _verify_virtctl_vms(self):
+            """
+            This method verifies the virtctl SSH login for each VM, either during the upgrade or once for each VM.
+            It prepares the data for ElasticSearch, generates a must-gather in case of an error, and uploads it to Google Drive.
+            """
+            try:
+                vm_names = self._oc._get_all_vm_names()
+                if not vm_names:
+                    raise MissingVMs("No VM names were retrieved from the cluster.")
 
-                # Error log with details of failed VM, for catching all vm errors
-                logger.error(
-                    f"Failed to verify virtctl SSH login for the following VMs: {', '.join(failure_vms)}")
+                upgrade_done = True
+                failure_vms = []  # List to store failed VM names
 
-        except Exception as err:
-            # Save run artifacts logs
-            self.save_error_logs()
-            raise err
+                if self._wait_for_upgrade_version:
+                    logger.info(f"wait for ocp upgrade version: {self._wait_for_upgrade_version}")
+                    upgrade_done = self._oc.get_cluster_status() == f'Cluster version is {self._wait_for_upgrade_version}'
+                    current_wait_time = 0
+
+                    while (self._timeout <= 0 or current_wait_time <= self._timeout) and not upgrade_done:
+                        failure_vms = self._verify_vms_in_parallel(vm_names)
+                        upgrade_done = self._oc.get_cluster_status() == f'Cluster version is {self._wait_for_upgrade_version}'
+
+                        # Sleep 1 sec between each cycle
+                        time.sleep(1)
+                        current_wait_time += 1  # Increment the wait time
```

Comment:
The verification probably takes a fair bit of time, so rather than assuming that the 1 second delay is the actual wait time,  get the starting time before starting the loop and compare the current time against the starting time for checking timeout.

Also, given that this is probably fairly resource-intensive, a 1 second interval between checks might be a bit too short and put unnecessary load on the server.

Reply:
That's still going to add up fast and if you're verifying 240 at once that will be a lot of load on the server.

Reply:
Ok, I’ve updated it to use a constant SLEEP_TIME of 3 seconds instead of 1.
BTW, we have a verification mechanism that handles each VM individually. 
I added this delay just to be on the safe side.